### PR TITLE
Combine Studio change for 2d components

### PR DESCRIPTION
**Pull request by xiaofeng11**
Move from https://github.com/cocos2d/cocos2d-x/pull/14372

### CODE REVIEW

**Comment by xiaofeng11** [comment]
@ricardoquesada @minggo Combine for studio changed 2d components move from https://github.com/cocos2d/cocos2d-x/pull/14372 to this PR. I have removed studio only camera define from this PR. Please review

**Comment by zilongshanren** [comment]
@xiaofeng11 This PR have merge conflicts, please resolve it.

**Comment by xiaofeng11** [comment]
@zilongshanren Fixed.

**Comment by xiaofeng11** [comment]
[ci rebuild]

File: tools/bindings-generator [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 250673608e23c9644b96a8e9ef886ab4db96aaf7
```
Comment (pandamicro) [inline tools/bindings-generator]: Do not change sub modules if not necessary

**Comment by xiaofeng11** [comment]
Add new api test in cpp-test project
Remove Sprite::setOffsetPosFromCenter function added by cocos studio, and reset _unflippedOffsetPositionFromCenter to Vec2::Zero when call Sprite::setTexture(const std::string &filename) function, Because when user load the sprite from a plist file, then load it from a picture file again, the _unflippedOffsetPositionFromCenter is still use plist setting, cause sprite render at wrong position.

@ricardoquesada @pandamicro Please review the change, Thanks.

File: plugin [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit ad539b56c5354c1083f479d4c8d275ae8c7fdfa7
```
Comment (pandamicro) [inline plugin]: This is still unchanged

**Comment by ricardoquesada** [comment]
`Sprite` and `Particles` should not include a CocosStudio file.  They are core components and they should not depend on external components.

**Comment by xiaofeng11** [comment]
@ricardoquesada CocosStudio header removed from Sprite and Particles, Please review, Thanks.

File: cocos/2d/CCDrawNode.cpp [hunk 3]
```diff
@@ -407,8 +417,16 @@ void DrawNode::onDrawGLLine(const Mat4 &transform, uint32_t flags)
     }
     
     glBindBuffer(GL_ARRAY_BUFFER, 0);
-    
     CC_INCREMENT_GL_DRAWN_BATCHES_AND_VERTICES(1,_bufferCountGLLine);
+
+#if CC_TARGET_PLATFORM == CC_PLATFORM_MAC || CC_TARGET_PLATFORM == CC_PLATFORM_WIN32
```
Comment (ricardoquesada) [inline cocos/2d/CCDrawNode.cpp]: why this? If this is a cocos studio thing, then the `#ifdef` should be `#if cocosstudio`

File: cocos/2d/CCSpriteFrameCache.cpp [hunk 4]
```diff
@@ -543,4 +548,177 @@ SpriteFrame* SpriteFrameCache::getSpriteFrameByName(const std::string& name)
...
+        texturePath = plist;
+
+        // remove .xxx
+        size_t startPos = texturePath.find_last_of(".");
+        texturePath = texturePath.erase(startPos);
+
+        // append .png
+        texturePath = texturePath.append(".png");
+
+        CCLOG("cocos2d: SpriteFrameCache: Trying to use file %s as texture", texturePath.c_str());
```
Comment (ricardoquesada) [inline cocos/2d/CCSpriteFrameCache.cpp]: please remove the CCLOG debug lines.

File: cocos/2d/CCTMXTiledMap.cpp [hunk 5]
```diff
@@ -29,6 +29,7 @@ THE SOFTWARE.
 #include "2d/CCTMXLayer.h"
 #include "2d/CCSprite.h"
 #include "deprecated/CCString.h" // For StringUtils::format
+#include "2d/CocosStudioExtension.h"
```
Comment (ricardoquesada) [inline cocos/2d/CCTMXTiledMap.cpp]: please remove this dependency as well

File: cocos/2d/CocosStudioExtension.cpp [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#include "CocosStudioExtension.h"
```
Comment (ricardoquesada) [inline cocos/2d/CocosStudioExtension.cpp]: why this file is created under the `2d` folder?  CocosStudio has its own folder.

File: plugin (hunk 2 above)
Comment (ricardoquesada) [inline plugin:1]: same comment as @pandamicro , why the plugin submodule was updated?

File: cocos/2d/CCDrawNode.cpp (hunk 3 above)
Comment (xiaofeng11) [inline cocos/2d/CCDrawNode.cpp]: GL_MULTISAMPLE and GL_LINE_SMOOTH only works on desktop system, GLES won't support these features, so it been limited to Windows and MacOS

File: plugin (hunk 2 above)
Comment (xiaofeng11) [inline plugin:1]: last plugin cocos2dx version is still 3.7, this commit is change version number to 3.9 in plugin.

**Comment by xiaofeng11** [comment]
@ricardoquesada Update code as comment, please review. Thanks.

File: cocos/2d/CCDrawNode.cpp [hunk 7]
```diff
@@ -398,6 +398,21 @@ void DrawNode::onDrawGLLine(const Mat4 &transform, uint32_t flags)
         // texcood
         glVertexAttribPointer(GLProgram::VERTEX_ATTRIB_TEX_COORD, 2, GL_FLOAT, GL_FALSE, sizeof(V2F_C4B_T2F), (GLvoid *)offsetof(V2F_C4B_T2F, texCoords));
     }
+
+#if CC_TARGET_PLATFORM == CC_PLATFORM_MAC || CC_TARGET_PLATFORM == CC_PLATFORM_WIN32
```
Comment (ricardoquesada) [inline cocos/2d/CCDrawNode.cpp]: the issue with this code is that we are adding functionality that only works on certain platforms.
for example, if I do `setLineSmooth(true)` on an iOS device, what will happen?
It won't work, but the user won't receive any kind of error.

Comment (xiaofeng11) [inline cocos/2d/CCDrawNode.cpp]: @ricardoquesada Could you give me some suggestion on this issue? Make setLineSmooth cocos studio only, or when user call it on a mobile platform give a error message, or other? Thank you very much!

Comment (ricardoquesada) [inline cocos/2d/CCDrawNode.cpp]: We shouldn't add features that don't work in all platforms.
My suggestion is that we temporary remove this feature from this PR until we find a better solution.

**Comment by xiaofeng11** [comment]
@ricardoquesada Desktop only function removed from this PR, and we set onDraw... functions in CCDrawNode override able, Please review, Thanks.

**Comment by ricardoquesada** [comment]
looks good to me.

File: cocos/2d/CCDrawNode.h [hunk 8]
```diff
@@ -298,23 +300,26 @@ class CC_DLL DrawNode : public Node
     /**
      * @js NA
      */
-	void onDraw(const Mat4 &transform, uint32_t flags);
+	virtual void onDraw(const Mat4 &transform, uint32_t flags);
```
Comment (pandamicro) [inline cocos/2d/CCDrawNode.h]: Indent issue, please use 4 spaces

File: cocos/2d/CCDrawNode.h [hunk 9]
```diff
@@ -298,23 +300,26 @@ class CC_DLL DrawNode : public Node
...
     /**
      * @js NA
      */
-	void onDrawGLLine(const Mat4 &transform, uint32_t flags);
+	virtual void onDrawGLLine(const Mat4 &transform, uint32_t flags);
```
Comment (pandamicro) [inline cocos/2d/CCDrawNode.h]: Indent issue, please use 4 spaces

**Comment by xiaofeng11** [comment]
@pandamicro indent fixed.

**Pull request merged**